### PR TITLE
Fix ambiguous documentation and implementation for eptri ep_out priming behaviour

### PR DESCRIPTION
**Pull request by antoinevg**
This PR is part of:
* https://github.com/greatscottgadgets/cynthion/pull/203
* https://github.com/greatscottgadgets/facedancer/pull/124

---

The documentation and implementation for eptri's `OutFIFOInterface` when it comes to endpoint priming has been somewhat at odds.

To whit, the `epno` register documentation read:

> Selects the endpoint number to prime. This interface only allows priming a single endpoint at once -- that is, only one endpoint can be ready to receive data at a time. See the `enable` bit for usage.

While the `prime` register documentation read:

> Controls "priming" an out endpoint. To receive data on any endpoint, the CPU must first select the endpoint with the `epno` register; and then write a '1' into the prime and enable register. This prepares our FIFO to receive data; and the next OUT transaction will be captured into the FIFO.
>
> When a transaction is complete, the `enable` bit is reset; the `prime` is not. This effectively means that `enable` controls receiving on _any_ of the primed endpoints; while `prime` can be used to build a collection of endpoints willing to participate in receipt.
>
> Only one transaction / data packet is captured per `enable` write; repeated enabling is necessary to capture multiple packets.

The actual implementation in this case was in line with the `epno` register in that endpoints would be un-primed upon receipt of a packet and would need to be re-primed before further data would be received:

```python
# If we've just ACK'd a receive, clear our enable, un-prime the given endpoint and
# clear our FIFO's ready state.
with m.If(interface.handshakes_out.ack & token.is_out):
    m.d.usb += [
        self.enable.r_data                .eq(0),
        endpoint_primed[token.endpoint]   .eq(0),  # <==== unprime endpoint
        fifo_ready                        .eq(0),
    ]
```

In practice, the behaviour documented in the `prime` register is far more desirable as it removes the need for the firmware to keep track of re-priming endpoints and only requires that the interface be re-enabled after receipt of a packet on any endpoint.

This PR removes the ambiguity from the documentation and establishes the behaviour documented in the `prime` register.

### CODE REVIEW

File: luna_soc/gateware/csr/usb2/interfaces/eptri.py [hunk 1]
```diff
@@ -642,6 +646,11 @@ def elaborate(self, platform):
                 self.enable.r_data                .eq(0),
                 fifo_ready                        .eq(0),
             ]
+            # If we've ACK'd a receive on the control endpoint, un-prime it to
+            # ensure we only receive control data _after_ we've had an opportunity
+            # to receive the setup packet.
+            with m.If(token.endpoint == 0):
+                m.d.usb += endpoint_primed[token.endpoint].eq(0)
```
Comment (martinling) [inline luna_soc/gateware/csr/usb2/interfaces/eptri.py:653]: Everything is working for me now, but I don't think this is the correct solution here.

What matters at this point is not really that the endpoint number is zero.

The issue as I see it, is that a lot of code in this module is concerned with whether the current transaction includes an outbound `DATAx` packet, and it's using `token.is_out` as the boolean indicator of that.

However, an `OUT` token is not the only thing that can be followed by an outbound `DATAx` packet. A `SETUP` token is also followed by an outbound `DATA` packet, and none of the code seems to have been accounting for that.

This change does make things work, as long as `SETUP` packets are only ever received on endpoint 0, which is *normally* the case. But although I've never seen it on the wire, the standard does allow devices to declare additional control endpoints. So you could get a `SETUP` for a non-zero endpoint, and then you'd be back to the same problem.

I suspect that what we really need to do here is to define something like:

`outbound_data = (token.is_out | token.is_setup)`

and then substitute `outbound_data` for `token.is_out` in the rest of the module.

Comment (antoinevg) [reply luna_soc/gateware/csr/usb2/interfaces/eptri.py:653]: In principle I 100% agree with you but in practice things get a little complicated due to the eptri architecture splitting setup packet handling into its own `SetupFIFOInterface`.  

i.e. The `token.is_setup` is case is being handled by `SetupFIFOInterface` which gets the `SETUP` packet data -- the `DATA` stage of the `SETUP` transaction is being handled by `OutFIFOInterface`.

You've put your finger on one of the biggest architectural problems with eptri which forces marshalling of `DATA` packets that are part of a `SETUP` transaction _vs_ `OUT` transactions into the _firmware_. 🥲

I'm not happy either about assuming zero for the control endpoint but this is also a problem that's pervasive throughout the Facedancer codebase and not one I think we should even try to solve within the limitations of eptri.

tl;dr -- eptri is an surprisingly powerful hack but I think it's going to be easier to replace eptri with a sane design than trying to make it do the things we take for granted in standard USB controller peripherals.

Comment (martinling) [reply luna_soc/gateware/csr/usb2/interfaces/eptri.py:653]: OK, that all makes sense.

I was looking at this fix in isolation and wasn't familiar with the bigger picture - I didn't know the background of the eptri design. With some more reading around I'm discovering that it's basically just a neat hack that lets you multiplex a whole multi-endpoint USB stack onto just three FIFOs, and I get the constraints associated with that a bit more now.

The fact that the device must always be ready to accept a `SETUP` is what makes that have to be three FIFOs rather than two, and at that point it makes sense that things are going to get messy like this.

So yes - let's go ahead and merge this and we can think about a fancier peripheral in the future.

**Review by martinling — APPROVED** [review]
